plugins/dynamix/include/OpenTerminal.php

```diff
@@ -0,0 +1,57 @@
+<?PHP
+/* Copyright 2005-2021, Lime Technology
+ * Copyright 2012-2021, Bergware International.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ */
+?>
+<?
+$docroot = $docroot ?? $_SERVER['DOCUMENT_ROOT'] ?: '/usr/local/emhttp';
+require_once "$docroot/webGui/include/Secure.php";
+
+switch ($_GET['tag']) {
+case 'ttyd':
+  $pid = exec("pgrep -a ttyd|awk '/\\/var\\/run\\/ttyd.sock:{print \$1}'");
+  if ($pid) break;
+  @unlink('/var/run/ttyd.sock');
+  exec("ttyd-exec -o -i '/var/run/ttyd.sock' bash --login");
+  break;
+case 'syslog':
+  $pid = exec("pgrep -a ttyd|awk '/\\/var\\/run\\/syslog.sock:{print \$1}'");
+  if ($pid) break;
+  $path = '/var/log/';
+  $file = realpath($path.$_GET['name']);
```

Comment:
Haven't run this but I have a feeling if I set `$_GET['name']` to `../../../etc/passwd` I'll get that file back. This needs a basename check to ensure the `$file` ends up in the `$path` after resolving.

Reply:
realpath and the subsequent path check takes care of this.